src/Ocelot/Configuration/Creator/RouteKeyCreator.cs

```diff
@@ -1,17 +1,58 @@
 using Ocelot.Configuration.File;
-using Ocelot.LoadBalancer.LoadBalancers;
+using Ocelot.LoadBalancer.LoadBalancers;
 
 namespace Ocelot.Configuration.Creator
 {
     public class RouteKeyCreator : IRouteKeyCreator
     {
-        public string Create(FileRoute fileRoute) => IsStickySession(fileRoute)
-            ? $"{nameof(CookieStickySessions)}:{fileRoute.LoadBalancerOptions.Key}"
-            : $"{fileRoute.UpstreamPathTemplate}|{string.Join(',', fileRoute.UpstreamHttpMethod)}|{string.Join(',', fileRoute.DownstreamHostAndPorts.Select(x => $"{x.Host}:{x.Port}"))}";
-
-        private static bool IsStickySession(FileRoute fileRoute) =>
-            !string.IsNullOrEmpty(fileRoute.LoadBalancerOptions.Type)
-            && !string.IsNullOrEmpty(fileRoute.LoadBalancerOptions.Key)
-            && fileRoute.LoadBalancerOptions.Type == nameof(CookieStickySessions);
+        public string Create(FileRoute fileRoute)
+        {
+            var isStickySession = fileRoute.LoadBalancerOptions is
+            {
+                Type: nameof(CookieStickySessions),
+                Key.Length: > 0
+            };
+
+            if (isStickySession)
+            {
+                return $"{nameof(CookieStickySessions)}:{fileRoute.LoadBalancerOptions.Key}";
+            }
+
+            var keyBuilder = new StringBuilder();
+            foreach (var upstreamHttpMethod in fileRoute.UpstreamHttpMethod)
+            {
+                Append(upstreamHttpMethod, ',');
+            }
+
+            Append(fileRoute.UpstreamHost);
+            Append(fileRoute.UpstreamPathTemplate);
+
+            if (!string.IsNullOrEmpty(fileRoute.ServiceName))
+            {
+                Append(fileRoute.ServiceNamespace);
+                Append(fileRoute.ServiceName);
+            }
+            else
+            {
+                foreach (var downstreamHostAndPorts in fileRoute.DownstreamHostAndPorts)
+                {
+                    Append(downstreamHostAndPorts.Host);
+                    Append(downstreamHostAndPorts.Port.ToString(), ':');
+                }
+            }
+
+            return keyBuilder.ToString();
+
+            // Helper method for appending a part to the key, automatically inserts a separator as needed
+            void Append(string next, char separator = '|')
+            {
+                if (keyBuilder.Length > 0)
+                {
+                    keyBuilder.Append(separator);
+                }
+
+                keyBuilder.Append(next);

```

Comment:
You don't analyze empty value of `next` argument!
So, when empty value then we add separator only. Is it right?

Reply:
Yea, I couldn't decide which is better, keeping or removing the empty parts. Keeping them felt like a safe choice, but I removed them now.